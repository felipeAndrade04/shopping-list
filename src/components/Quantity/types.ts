/* eslint-disable no-unused-vars */
export interface QuantityProps {
  initialValue?: number;
  value?: number;
  onChange?: (value: number) => void;
}
