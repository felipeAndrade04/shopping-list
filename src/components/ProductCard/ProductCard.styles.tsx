import styled from 'styled-components/native';

export const Product = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  padding: 16px;
  border-radius: 4px;
  background: ${({ theme }) => theme.colors.white};
`;

export const LeftProductInfo = styled.View`
  display: flex;
  flex-direction: row;
`;

export const ProductName = styled.Text`
  font-size: 16px;
  font-family: ${({ theme }) => theme.fonts.medium};
`;
