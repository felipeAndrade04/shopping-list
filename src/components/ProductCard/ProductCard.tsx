import { Checkbox } from '@app/components/Checkbox';
import { Quantity } from '@app/components/Quantity';
import { Spacer } from '@app/components/Spacer';
import { useState } from 'react';
import * as S from './ProductCard.styles';
import { ProductCardProps } from './ProductCard.types';

export function ProductCard({
  product,
  isChecked: isCheckedProp,
  changeQuantity,
  changeSelectedProduct,
}: ProductCardProps) {
  const [isChecked, setIsChecked] = useState(isCheckedProp);
  const [quantity, setQuantity] = useState(product?.quantity);

  function onChangeQuantity(quantity: number) {
    changeQuantity(product.id, quantity);
    setQuantity(quantity);
  }

  function onChangeCheckbox(value: boolean) {
    if (!value) {
      setQuantity(1);
    }
    changeSelectedProduct(product, value);
    setIsChecked(value);
  }

  return (
    <S.Product>
      <S.LeftProductInfo>
        <Checkbox value={isChecked} onChange={onChangeCheckbox} />
        <Spacer dimesion={12} />
        <S.ProductName>{product.name}</S.ProductName>
      </S.LeftProductInfo>

      {isChecked && <Quantity value={quantity} onChange={onChangeQuantity} />}
    </S.Product>
  );
}
