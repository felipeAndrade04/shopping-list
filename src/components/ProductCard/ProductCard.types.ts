/* eslint-disable no-unused-vars */
import { Product } from '@app/models';

export interface ProductCardProps {
  product: Product;
  isChecked: boolean;
  changeSelectedProduct?: (product: Product, value: boolean) => void;
  changeQuantity?: (productId: string, quantity: number) => void;
}
