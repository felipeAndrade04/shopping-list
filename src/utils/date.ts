const months = ['jan', 'fev', 'mar', 'abr', 'mai', 'jun', 'jul', 'ago', 'set', 'out', 'nov', 'dez'];

export function formatDate(date: Date) {
  const day = date.getDay();
  const month = date.getMonth();

  return `${day} de ${months[month]}`;
}
