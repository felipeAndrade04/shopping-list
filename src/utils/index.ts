export * from './delay';
export * from './date';
