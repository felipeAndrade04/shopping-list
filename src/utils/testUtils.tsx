import React from 'react';
import { authReducer } from '@app/store';
import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components/native';
import * as theme from '@app/theme';
import { configureStore } from '@reduxjs/toolkit';

let store = configureStore({
  reducer: {
    auth: authReducer,
  },
});

export const wrapper = ({ children }: { children: ReactNode }) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

export function resetStore() {
  store = configureStore({
    reducer: {
      auth: authReducer,
    },
  });
}
