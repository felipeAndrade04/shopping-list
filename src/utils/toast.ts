import EventManager from './EvenManager';

export type ToastType = 'success' | 'info' | 'error';

export interface Toast {
  id?: string;
  type: ToastType;
  text: string;
  duration?: number;
}

export const toastEventManager = new EventManager();

export default function toast({ type, text, duration }: Toast) {
  toastEventManager.emit<Toast>('addtoast', {
    id: new Date().getTime().toString(),
    type,
    text,
    duration,
  });
}
