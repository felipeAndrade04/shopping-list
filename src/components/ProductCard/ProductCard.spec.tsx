import { render } from '@testing-library/react-native';
import { wrapper } from '@app/utils/testUtils';
import { ProductCard } from './ProductCard';
import { ProductType } from './ProductCard.types';

const product: ProductType = {
  id: '1',
  name: 'Sab√£o',
  category: 'Limpeza',
  checked: true,
};

describe('<ProductCard />', () => {
  it('Should be render correctly', () => {
    const tree = render(<ProductCard isChecked={true} product={product} />, { wrapper }).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should be render unchecked product', () => {
    const { queryByTestId, getByText } = render(
      <ProductCard isChecked={false} product={{ ...product, checked: false }} />,
      { wrapper }
    );

    expect(queryByTestId('checkbox-icon')).toBeFalsy();
    expect(getByText(product.name)).toBeTruthy();
    expect(queryByTestId('quantity')).toBeFalsy();
  });

  it('Should be render checked product', () => {
    const { queryByTestId, getByText } = render(
      <ProductCard isChecked={true} product={product} />,
      { wrapper }
    );

    expect(queryByTestId('checkbox-icon')).toBeTruthy();
    expect(getByText(product.name)).toBeTruthy();
    expect(queryByTestId('quantity')).toBeTruthy();
  });
});
