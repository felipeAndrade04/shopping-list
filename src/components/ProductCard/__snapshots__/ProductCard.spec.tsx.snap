// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> Should be render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 4,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#009Cf7",
          "borderColor": "#009Cf7",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 2,
          "display": "flex",
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 20,
        }
      }
      testID="checkbox"
    >
      <View
        testID="checkbox-icon"
      >
        <Text />
      </View>
    </View>
    <View
      dimesion={12}
      style={
        [
          {
            "height": 12,
            "width": 12,
          },
        ]
      }
    />
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Medium",
            "fontSize": 16,
          },
        ]
      }
    >
      Sab√£o
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
    testID="quantity"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#009Cf7",
          "borderRadius": 4,
          "display": "flex",
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
        }
      }
      testID="decrement-quantity"
    >
      <Text />
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Medium",
            "fontSize": 16,
            "paddingBottom": 0,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 0,
          },
        ]
      }
    >
      1
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#009Cf7",
          "borderRadius": 4,
          "display": "flex",
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
        }
      }
      testID="increment-quantity"
    >
      <Text />
    </View>
  </View>
</View>
`;
